eeps/eep-0057.md

```diff
@@ -0,0 +1,400 @@
+in guards, "if" statements, and list comprehensions.
+
+However, the semicolon (`;`) could create ambiguity in identifying
+the end of the pattern in a statement or in separating expressions. E.g.:
+```
+    case Expr of
+      a ->
+        true = foo();  % <-- is this the end of the body or a syntax error
+                       %     which should be a comma, with the body
+                       %     returning 'b'?
+
+      b;               % <-- is this the alternative pattern or a pattern
+                       %     with a missing body?
+      C when is_integer(C) ->
+        false
+    end
+
+    % Does `1; 2` here represent two OR'd expressions or an alternative
+    % pattern?
+    if 1; 2 = length(L) ->
+      true;
+    true ->
+      false
+    end
+
+```
+
+
+Nested Alternative Patterns
+---------------------------
+
+There are two approaches of dealing with the alternative patters:
+
+1. Permitting only the top-level alternation (e.g.
+   `Expr1 | Expr2 | ... | ExprN`)
+2. Permitting unlimited nesting
+   (e.g. `{a|{b|c, d}} | e | {f|g, [(h|i), (j|k) | [(l|m)]]}`)
+
+While supporting the nested patterns is desirable, depending on the
+complexity of implementation, the unlimited nesting may be decided
+not to be supported.  More specifically, it may turn out that the
+complexity of implementing nested patterns by the compiler outweighs
+the benefits of this feature.
+
+
+
+Overlapping Patterns
+--------------------
+
+In cases when alternative patterns overlap, the natural left-to-right
+evaluation order would be used similar to the order of patterns in
+a `case` statement. E.g.
+```
+    {a, X} | {X, b} = foo()
+```
+This example would be equivalent to:
+```
+    case foo() of
+      {a, X} -> ...
+      {X, b} -> ...
+    end
+```
+
+Rationale
+=========
+
+Implementation of this proposal would lead to a more
+convenient way to code redundant matching clauses, and
+would eliminate code duplication.
+
+Example of where this could be useful can be found in
+many places in the existing OTP code. E.g. ssl_logger.erl:
+```
+    case logger:compare_levels(Level, debug) of
+        lt ->
+            ?LOG_DEBUG(#{direction => Direction,
+                         protocol => Protocol,
+                         message => Message},
+                       #{domain => [otp,ssl,Protocol]});
+        eq ->
+            ?LOG_DEBUG(#{direction => Direction,
+                         protocol => Protocol,
+                         message => Message},
+                       #{domain => [otp,ssl,Protocol]});
+        _ ->
+            ok
+    end.
+
+    case logger:compare_levels(Level, notice) of
+        lt ->
+            ?LOG_NOTICE(Report);
+        eq ->
+            ?LOG_NOTICE(Report);
+        _ ->
+            ok
+    end.
+```
+
+Additionally, this syntax extension leads to consistency between
+function spec definitions and their implementations:
+```
+    -spec f(foo|bar) -> ok.
+    f(foo|bar) -> ok.
```

Comment:
That's a very strong example in favor of using `|`. Perhaps it should be moved into `Choice of a Delimiter for Alternative Patterns`.

Reply:
You mean, the example itself?  Because this reason it's already stated there on line 209.

Reply:
Yes the example. This example is worth a thousand words.